File: src/main/java/me/ruslanys/vkmusic/javafx/controller/MainController.java
package me.ruslanys.vkmusic.javafx.controller;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.control.SelectionMode;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import me.ruslanys.vkmusic.annotation.FxmlController;
import me.ruslanys.vkmusic.component.VkClient;
import me.ruslanys.vkmusic.domain.Audio;
import me.ruslanys.vkmusic.util.IconUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;

@FxmlController(view = "views/fxml/main.fxml")
public class MainController {

    @FXML public Pane loadingView;
    @FXML public ImageView loadingImageView;
    @FXML private TableView<Audio> tableView;

    @Autowired
    private VkClient vkClient;


    private ObservableList<Audio> data;


    @FXML
    public void initialize() {
        initLoadingIcon();
        initTableView();
    }

    private void initLoadingIcon() {
        loadingImageView.setImage(IconUtils.INSTANCE.getLoadingIcon());
    }

    private void initTableView() {
        data = FXCollections.observableArrayList(new ArrayList<>());

        // Столбцы таблицы
        TableColumn<Audio, String> idColumn = new TableColumn<>("ID");
        idColumn.setCellValueFactory(new PropertyValueFactory<>("id"));

        TableColumn<Audio, String> artistColumn = new TableColumn<>("Исполнитель");
        artistColumn.setCellValueFactory(new PropertyValueFactory<>("artist"));

        TableColumn<Audio, String> titleColumn = new TableColumn<>("Наименование");
        titleColumn.setCellValueFactory(new PropertyValueFactory<>("title"));

        TableColumn<Audio, Integer> durationColumn = new TableColumn<>("Продолжительность");
        durationColumn.setCellValueFactory(new PropertyValueFactory<>("duration"));

        TableColumn<Audio, String> statusColumn = new TableColumn<>("Статус");
        statusColumn.setCellValueFactory(new PropertyValueFactory<>("status"));

        tableView.getColumns()
                .setAll(idColumn, artistColumn, titleColumn, durationColumn, statusColumn);

        tableView.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);

        // Данные таблицы
        tableView.setItems(data);
    }

    @FXML
    public void selectAll() {
        tableView.getSelectionModel().selectAll();
    }

    @FXML
    public void refreshTable() {
        tableView.setVisible(false);

        CompletableFuture
                .supplyAsync(() -> vkClient.getAudio(vkClient.fetchUserId()))
                .thenAccept(audio -> {
                    data.clear();
                    data.addAll(audio);
                    tableView.setVisible(true);
                });
    }

}

File: src/main/kotlin/me/ruslanys/vkmusic/ui/controller/MainController.kt
package me.ruslanys.vkmusic.ui.controller

import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.fxml.FXML
import javafx.scene.control.SelectionMode
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.cell.PropertyValueFactory
import javafx.scene.image.ImageView
import javafx.scene.layout.Pane
import me.ruslanys.vkmusic.annotation.FxmlController
import me.ruslanys.vkmusic.component.VkClient
import me.ruslanys.vkmusic.domain.Audio
import me.ruslanys.vkmusic.util.IconUtils
import java.util.concurrent.CompletableFuture

@FxmlController(view = "views/main.fxml")
class MainController(private val vkClient: VkClient) : BaseController() {

    @FXML private lateinit var loadingView: Pane
    @FXML private lateinit var loadingImageView: ImageView
    @FXML private lateinit var tableView: TableView<Audio>

    private lateinit var data: ObservableList<Audio>

    @FXML
    fun initialize() {
        initLoading()
        initTable()
    }

    private fun initLoading() {
        loadingImageView.image = IconUtils.getLoadingIcon()
    }

    private fun initTable() {
        data = FXCollections.observableArrayList(arrayListOf())

        // Columns
        val idColumn = TableColumn<Audio, String>("ID").also {
            it.setCellValueFactory(PropertyValueFactory<Audio, String>("id"))
        }

        val artistColumn = TableColumn<Audio, String>("Исполнитель").also {
            it.setCellValueFactory(PropertyValueFactory<Audio, String>("artist"))
        }

        val titleColumn = TableColumn<Audio, String>("Наименование").also {
            it.setCellValueFactory(PropertyValueFactory<Audio, String>("title"))
        }

        val durationColumn = TableColumn<Audio, String>("Продолжительность").also {
            it.setCellValueFactory(PropertyValueFactory<Audio, String>("duration"))
        }

        val statusColumn = TableColumn<Audio, String>("Статус").also {
            it.setCellValueFactory(PropertyValueFactory<Audio, String>("status"))
        }

        // --
        tableView.columns.setAll(idColumn, artistColumn, titleColumn, durationColumn, statusColumn)
        tableView.selectionModel.selectionMode = SelectionMode.MULTIPLE

        // Data
        tableView.items = data
    }

    @FXML
    fun selectAll() {
        tableView.selectionModel.selectAll()
    }

    @FXML
    fun refreshTable() {
        tableView.isVisible = false

        CompletableFuture.supplyAsync {
            vkClient.getAudio()
        }.thenAccept {
            data.clear()
            data.addAll(it)
            tableView.isVisible = true
        }
    }

}